T7424: Refactor resource validation and broaden cases

oniko94 opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [x] Other (please describe):

Refactoring the resource usage (CPU and memory) checks for VPP configuration mode on verify stage for more fine-grained resource utilization when configuring.

## Related Task(s)
<!-- optional: Link to related other tasks on Phabricator. -->
<!-- * https://vyos.dev/Txxxx -->
https://vyos.dev/T7424
## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->

## Proposed changes
<!--- Describe your changes in detail -->

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

oniko94 committed:
T7424: Refactor and extend resource usage verification on commit for VPP CLI

T7424: Fix ruff errors

sever-sever commented:
Add the task number on the https://vyos.dev/ and refer the task in the commit message and PR message

sever-sever commented on python/vyos/vpp/config_resource_checks/cpu.py:
```diff
@@ -0,0 +1,94 @@
+from vyos.vpp.config_resource_checks.constants import RESERVED_CPU_CORES
+
+
+def __total_cpu_count() -> int:
+    return int(popen("egrep '^core id' /proc/cpuinfo | sort -u | wc -l").read())
```
We do not use `grep/egrep`  for Python scripts
But even if I use this grep, I got incorrect values for the system with 4 CPU's
```
vyos@r14:~$ egrep '^core id' /proc/cpuinfo | sort -u | wc -l
1
vyos@r14:~$ 

```
check:
```
vyos@r14:~$ lscpu
Architecture:             x86_64
  CPU op-mode(s):         32-bit, 64-bit
  Address sizes:          48 bits physical, 48 bits virtual
  Byte Order:             Little Endian
CPU(s):                   4
  On-line CPU(s) list:    0-3
Vendor ID:                AuthenticAMD
  Model name:             AMD Ryzen 5 3600 6-Core Processor
    CPU family:           23
    Model:                113
    Thread(s) per core:   1
    Core(s) per socket:   1
    Socket(s):            4
    Stepping:             0
    BogoMIPS:             7186.49
    Flags:                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm rep_good no
                          pl cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor
                           lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw perfctr_core ssbd ibpb stibp vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 r
                          dseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 clzero xsaveerptr wbnoinvd arat npt lbrv nrip_save tsc_scale vmcb_clean flushbyasid pausefil
                          ter pfthreshold v_vmsave_vmload vgif umip rdpid arch_capabilities
Virtualization features:  
  Virtualization:         AMD-V
  Hypervisor vendor:      KVM
  Virtualization type:    full
Caches (sum of all):      
  L1d:                    256 KiB (4 instances)
  L1i:                    256 KiB (4 instances)
  L2:                     2 MiB (4 instances)
  L3:                     64 MiB (4 instances)
NUMA:                     
  NUMA node(s):           1
  NUMA node0 CPU(s):      0-3

```

sever-sever commented on python/vyos/vpp/config_resource_checks/cpu.py:
```diff
@@ -0,0 +1,94 @@
+
+def __core_threads() -> int:
+    # Retrieve how many threads can a single physical CPU core have
+    stepping = (popen('egrep "^stepping" /proc/cpuinfo | sort -u | tr -d -c 0-9')
```
We do not use `grep/egrep` for Python scripts

sever-sever reviewed (changes requested):
There should be a task number on the https://vyos.dev/ and the task should be in the commit message and PR titile

oniko94 commented:
> There should be a task number on the https://vyos.dev/ and the task should be in the commit message and PR titile

@sever-sever done and linked

mykolabas commented on python/vyos/vpp/config_resource_checks/memory.py:
```diff
@@ -0,0 +1,122 @@
+        memory_effective = total_memory_required(config['effective'])
+        memory_used = memory_used - memory_effective
+
+    if memory_total >= (constants.RESERVED_MEM * 2):
```
I think, this check is irrelevant. memory_total should be subtracted by RESERVED_MEM without any conditions

sever-sever commented on python/vyos/vpp/config_resource_checks/memory.py:
```diff
@@ -0,0 +1,122 @@
+def total_memory_required(settings: dict) -> int:
+    mem_required = hugepage_size(settings)
+    mem_stats = {
+        'memory_buffers': buffer_size(settings, available_core_count()),
```
Where is `available_core_count` imported?

sever-sever commented on python/vyos/vpp/config_resource_checks/memory.py:
```diff
@@ -0,0 +1,122 @@
+import psutil
+
+from vyos.vpp.config_resource_checks import constants
+from vyos.vpp.config_resource_checks.cpu import core_count
```
Is it used in this file?

sever-sever commented:
Black does not like the style
```
1s
Run black --check --skip-string-normalization $(git ls-files '*.py')
  black --check --skip-string-normalization $(git ls-files '*.py')
  shell: /usr/bin/bash -e {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.11.12/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.11.12/x64/lib
would reformat python/vyos/vpp/config_resource_checks/memory.py
would reformat python/vyos/vpp/config_verify.py
```
Change please

oniko94 committed:
T7424: Implement check for smoke tests runtime; reduce resource requirements for test environments

T7424: Fix errors in calculating the skipped and reserved CPU cores; Adjust default main heap size value.

oniko94 committed:
T7424: Refactor the CPU checks logic; Add total CPU usage check

T7424: Fix CPU reserve and skip cores calculations; Add total CPU usage check

T7424: Refactor smoketests to reflect new logic

mykolabas commented on python/vyos/vpp/config_resource_checks/memory.py:
```diff
@@ -0,0 +1,133 @@
+
+
+def total_memory_required(settings: dict, defaults: dict) -> int:
+    mem_required = hugepage_size(settings, defaults.get('hugepage_size'))
```
it should be: mem_required = 0
hugepage_size isn't related to memory required for VPP

sever-sever commented:
@oniko94, why was it closed?
